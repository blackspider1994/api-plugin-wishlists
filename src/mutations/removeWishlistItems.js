import SimpleSchema from "simpl-schema";
import ReactionError from "@reactioncommerce/reaction-error";

const inputSchema = new SimpleSchema({
  "wishlistId": String,
  "wishlistItemIds": {
    type: Array,
    minCount: 1
  },
  "wishlistItemIds.$": String,
  "wishlistToken": {
    type: String,
    optional: true
  }
});

/**
 * @method removeWishlistItems
 * @summary Removes one or more items from a wishlist
 * @param {Object} context -  an object containing the per-request state
 * @param {Object} input - Necessary input
 * @param {String} input.wishlistId - The ID of the wishlist in which all of the items exist
 * @param {String[]} input.wishlistItemIds - Array of wishlist item IDs to remove
 * @param {String} input.wishlistToken - The wishlistToken if the wishlist is an anonymous wishlist
 * @returns {Promise<Object>} An object containing the updated wishlist in a `wishlist` property
 */
export default async function removeWishlistItems(context, input) {
  inputSchema.validate(input || {});

  const { accountId, collections } = context;
  const { Wishlist } = collections;
  const { wishlistId, wishlistItemIds } = input;

  const selector = { _id: wishlistId };
  if (accountId) {
    selector.accountId = accountId;
  } else {
    throw new ReactionError("invalid-param", "A wishlistToken is required when updating an anonymous wishlist");
  }

  const wishlist = await Wishlist.findOne(selector);
  if (!wishlist) throw new ReactionError("not-found", "Wishlist not found");

  const updatedWishlist = {
    ...wishlist,
    items: wishlist.items.filter((item) => !wishlistItemIds.includes(item._id)),
    updatedAt: new Date()
  };

  const savedWishlist = await context.mutations.saveWishlist(context, updatedWishlist);

  return { wishlist: savedWishlist };
}
