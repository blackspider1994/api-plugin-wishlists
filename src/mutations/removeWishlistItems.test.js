import mockContext from "@reactioncommerce/api-utils/tests/mockContext.js";
import removeWishlistItems from "./removeWishlistItems.js";

const dbWishlist = {
  _id: "wishlistId",
  items: []
};

const wishlistItemIds = ["wishlistItemId1", "wishlistItemId2"];

beforeAll(() => {
  if (!mockContext.mutations.saveWishlist) {
    mockContext.mutations.saveWishlist = jest.fn().mockName("context.mutations.saveWishlist").mockImplementation(async (_, wishlist) => wishlist);
  }

  mockContext.collections.Wishlist = {
    findOne: jest.fn().mockName("Wishlist.findOne")
  };
});

test("removes multiple items from account wishlist", async () => {
  mockContext.collections.Wishlist.findOne.mockReturnValueOnce(Promise.resolve(dbWishlist));

  const result = await removeWishlistItems(mockContext, { wishlistId: "wishlistId", wishlistItemIds });

  expect(mockContext.collections.Wishlist.findOne).toHaveBeenCalledWith({
    _id: "wishlistId",
    accountId: "FAKE_ACCOUNT_ID"
  });

  expect(result).toEqual({
    wishlist: {
      ...dbWishlist,
      items: dbWishlist.items.filter((item) => !wishlistItemIds.includes(item._id)),
      updatedAt: jasmine.any(Date)
    }
  });
});

test("throws when no account and no wishlistToken passed", async () => {
  const cachedAccountId = mockContext.accountId;
  mockContext.accountId = null;

  await expect(removeWishlistItems(mockContext, {
    wishlistId: "wishlistId",
    wishlistItemIds
  })).rejects.toThrowErrorMatchingSnapshot();

  mockContext.accountId = cachedAccountId;
});
