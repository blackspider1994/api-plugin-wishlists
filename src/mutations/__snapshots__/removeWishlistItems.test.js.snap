// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws when no account and no wishlistToken passed 1`] = `"A wishlistToken is required when updating an anonymous wishlist"`;
